src/components/shared/thread/ActivityGraphFills.js

```diff
@@ -685,32 +685,67 @@ export class ActivityFillGraphQuerier {
     const pixelsAroundX = new Float32Array(SMOOTHING_KERNEL.length);
     const sampleTime = samples.time[sample];
-    // xPixel in graph space maps to kernel[smoothingRadius]
-    const sampleTimeRangeStart =
-      sample > 0 ? (samples.time[sample - 1] + sampleTime) / 2 : -Infinity;
-    const sampleTimeRangeEnd =
-      sample + 1 < samples.length
-        ? (samples.time[sample + 1] + sampleTime) / 2
-        : Infinity;
-
-    let pixelStart =
-      (sampleTimeRangeStart - rangeStart) * xPixelsPerMs - kernelPos;
-    let pixelEnd = (sampleTimeRangeEnd - rangeStart) * xPixelsPerMs - kernelPos;
-    pixelStart = clamp(pixelStart, 0, SMOOTHING_KERNEL.length - 1);
-    pixelEnd = clamp(pixelEnd, 0, SMOOTHING_KERNEL.length - 1);
-    const intPixelStart = pixelStart | 0;
-    const intPixelEnd = pixelEnd | 0;
-
-    for (let i = intPixelStart; i <= intPixelEnd; i++) {
-      pixelsAroundX[i] += 1;
+    // Use the fullThread here to properly get the next and previous in case zoomed in.
+    const fullThreadSample = sample + sampleIndexOffset;
+    const prevSampleTime =
+      fullThreadSample > 0
+        ? fullThread.samples.time[fullThreadSample - 1]
+        : sampleTime - interval;
+    const nextSampleTime =
+      fullThreadSample + 1 < fullThread.samples.length
+        ? fullThread.samples.time[fullThreadSample + 1]
+        : sampleTime + interval;
+
+    let cpuBeforeSample = null;
+    let cpuAfterSample = null;
+    const { threadCPUDelta } = samples;
+    if (enableCPUUsage && threadCPUDelta) {
+      // It must be non-null because we are checking this in the processing
+      // step and eliminating all the null values.
+      const cpuDeltaBefore = ensureExists(threadCPUDelta[sample]);
+      // Use the fullThread here to properly get the next in case zoomed in.
+      const cpuDeltaAfter = ensureExists(fullThread.samples.threadCPUDelta)[
+        fullThreadSample + 1
+      ];
+      const intervalDistribution = (sampleTime - prevSampleTime) / interval;
+      const nextIntervalDistribution = (nextSampleTime - sampleTime) / interval;
+
+      // Figure out the CPU usage "per interval". This is needed for cases
+      // where we have some missing samples. For example:
+      //
+      // Interval:        1ms                  2ms                 1ms
+      // CPU:          100 cycles           200 cycles          200 cycles
+      // Samples:  [x-------------x--------------------------x-------------x]
+      //
+      // In this case, even though it has the max CPU cycle count, the CPU
+      // usage should be 50% because this cycle count is from 2ms area instead
+      // of 1ms like the latter one.
+      cpuBeforeSample = cpuDeltaBefore / intervalDistribution;
+      cpuAfterSample =
+        // It can be undefined if this is the last sample in the profile.
+        cpuDeltaAfter !== undefined && cpuDeltaAfter !== null
+          ? cpuDeltaAfter / nextIntervalDistribution
+          : cpuBeforeSample;
```

Comment:
While re-reading this I remembered that I was unhappy about the existing code here :) But since this is just a copy of something we already do, that's fine. We can improve it in a separate patch.

My main complaint is that we unnecessarily have the interval in various calculations. maxThreadCPUDelta has an interval factor inside of it; if we remove that we can also remove the multiplication and division with the interval in a bunch of other places. maxThreadCPUDelta should be per millisecond, not per interval. And ideally, the derived `threadCPUDelta` arrays should be per-millisecond ratios, not deltas. But, as I said, that should be a follow-up.

Reply:
Ah yeah, I remember that discussion. I think that was going to be a follow-up, but never ended up doing it 😄 Thanks for reminding it. I will do it as a follow-up.